x-pack/plugins/ml/public/util/app_state_utils.js

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+
+import { distinctUntilChanged } from 'rxjs/operators';
+
+export function initializeAppState(AppState, stateName, defaultState) {
+  const appState = new AppState();
+  appState.fetch();
+
+  // Store the state to the AppState so that it's
+  // restored on page refresh.
+  if (appState[stateName] === undefined) {
+    appState[stateName] = _.cloneDeep(defaultState) || {};
+    appState.save();
+  }
+
+  // If defaultState is defined, check if the keys of the defaultState
+  // match the one from appState, if not, fall back to the defaultState.
+  // If we didn't do this, the structure of an out-of-date appState
+  // might break some follow up code. Note that this will not catch any
+  // deeper nested inconsistencies.
+  if (typeof defaultState !== 'undefined' && appState[stateName] !== defaultState) {
+    // if defaultState is an object, check if current appState has the same keys.
+    // if it's not an object, check if defaultState and current appState are of the same type.
+    if (
+      (typeof defaultState === 'object' && !_.isEqual(
```

Comment:
could this `typeof` check be moved to the outer `if`, to replace the `undefined` check?

Reply:
I'm afraid this wouldn't work as expected: The most outer `if` checks if `defaultState` is set at all, the inner `if` contains two checks to find out if the data stored in AppState is valid:

- The one you commented on is about when `defaultState` is an object, then it gets compared to `appStates` version to see if they share the same object attributes. This is like duck typing the AppState object to find out if it has the same format then `defaultState`.
- The second check is when the state is simpler, like for `showCharts` it's just a boolean, then we compare the types of `defaultState` and `AppState`. For example, if `defaultState` for `showCharts` is `true` but for whatever reason the URL is malformed and `AppState` returns `123` it would fall back to `defaultState`.

By allowing these simpler values in AppState (`stateFactory()` always required an object) our URLs got a bit shorter, for example the value stored in the URL for `showCharts` now looks like this: `mlCheckboxShowCharts:!t`, previously it was something like `mlCheckboxShowCharts:(showCharts:!t)`